[CARBONDATA-2420][32K] Support string longer than 32000 characters

xuchuanyin opened the pull request:
Add a property in creating table 'long_string_columns' to support string columns that will contains more than 32000 characters.
Inside carbondata, it use an integer instead of short to store the length of bytes content.

Internally in Carbondata,
1. add a new datatype called `varchar` to represent the long string column
2. add a new encoding called `DIRECT_COMPRESS_VARCHAR` to the archer column page meta
3. use an integer (previously short) to store the length of bytes content.
4. add 2GB constraint for one column page

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 `Only internal interfaces have been changed`
 - [x] Any backward compatibility impacted?
 `NO`
 - [x] Document update required?
`YES`
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
`Added tests`
        - How it is tested? Please attach test report.
`Tested in local machine`
        - Is it a performance related change? Please attach the performance test report.
`NO`
        - Any additional information to help reviewers in testing this change.
      `NA` 
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
`NA`

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5546/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4383/

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4641/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4642/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5564/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4402/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4660/

xuchuanyin commented:
@ravipesala I've considered to add a datatype such as TEXT, but quit the idea due to that the grammar is not general, at least it is not compatible with Spark/Hive. It will cause problem to migrate from/to Carbondata.

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5567/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4664/

xuchuanyin commented:
@kumarvishal09 yeah, it’s an option to add an encoder type and make L-V related class abstract to eliminate the duplicate code.

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5025/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4859/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5027/

xuchuanyin commented:
To eliminate if-else on longString as much as possible,
+ For DataMapSchemaType, I changed VARIABLE to VARIABLE_INT and VARIABLE_SHORT, they are used for BlockletDataMap;
+ For DimensionStoreType, I changed VARIABLELENGTH to VARIABLE_INT_LENGTH and VARIABLE_SHORT_LENGTH, they are used for encoding/decoding dimensions;
+ For ColumnPageStatCollector, I changed LVStringStatCollector to LVShortStringStatCollector and LVLongStringStatCollector, they are used for column statistics;

While deep into the code, I found that there is no need to add an internal datatype such as TEXT.
+ The dimensions all are considered as String, we only need a boolean array to indicate whether it is long string. We don't need an array to indicate all the datatype of the dimensions.
+ The procedure for String and Text are nearly the same. A boolean (array) and proper abstraction is enough.

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5039/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6026/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4867/

kumarvishal09 commented on core/src/main/java/org/apache/carbondata/core/datastore/chunk/impl/DimensionRawColumnChunk.java:
```diff
@@ -92,8 +93,10 @@ public DimensionColumnPage decodeColumnPage(int pageNumber) {
    */
-  public DimensionColumnPage convertToDimColDataChunkWithOutCache(int index) {
+  public DimensionColumnPage convertToDimColDataChunkWithOutCache(int index,
+      boolean isLongStringColumn) {
```
why this is required???...as mentioned in earlier comment...reader does not know which type of data it is reading

xuchuanyin replied:
What does `reader` mean?

kumarvishal09 commented on core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v1/CompressedDimensionChunkFileBasedReaderV1.java:
```diff
@@ -99,6 +99,7 @@ public CompressedDimensionChunkFileBasedReaderV1(final BlockletInfo blockletInfo
 
   @Override public DimensionColumnPage decodeColumnPage(
       DimensionRawColumnChunk dimensionRawColumnChunk, int pageNumber) throws IOException {
+    boolean isLongStringColumn = dimensionRawColumnChunk.isLongStringColumn();
```
In V1 case it will be always false...as new encoder type will be only supported for V3 format

xuchuanyin replied:
Should we support longStringColumn in V1?

Carbondata now only support writing V1 format.

xuchuanyin replied:
I know, you mean just leave it false?

kumarvishal09 replied:
No we support V3 while writing the data...v1/v2 is supported only for reading to support backward compatibility

kumarvishal09 commented on core/src/main/java/org/apache/carbondata/core/datastore/chunk/reader/dimension/v2/CompressedDimensionChunkFileBasedReaderV2.java:
```diff
@@ -121,6 +121,7 @@ public DimensionColumnPage decodeColumnPage(
     int[] invertedIndexesReverse = new int[0];
     int[] rlePage = null;
     DataChunk2 dimensionColumnChunk = null;
+    boolean isLongStringColumn = dimensionRawColumnChunk.isLongStringColumn();
```
In V2 case it will be always false...as new encoder type will be only supported for V3 format

xuchuanyin replied:
The same as above. V2 does not support longStringColumn, so it is fine that the `isLongStringColumn` will always be false.

xuchuanyin replied:
The same as above

xuchuanyin commented:
@kumarvishal09 You suggested to add a new TEXT encoder and using this encoder while writing&reading.
But currently in CarbonData, all dimensions are considered as string, there is no specified encoder for it.
In the previous description,
```
For DimensionStoreType, I changed VARIABLELENGTH to VARIABLE_INT_LENGTH and VARIABLE_SHORT_LENGTH, they are used for encoding/decoding dimensions
```
We can consider the DimensionStoreType as an encoder. It had two values： FIXEDLENGTH and VARIABLELENGTH and I extended it to tree：FIXED_LENGTH、VARIABLE_SHORT_LENGTH、VARIABLE_INT_LENGTH.

Does this meet your suggestion?

kumarvishal09 commented:
@xuchuanyin how u are deciding isLongStringColumn is true or false in query??

xuchuanyin commented on core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentProperties.java:
```diff
@@ -849,7 +852,41 @@ public int getNumberOfDictSortColumns() {
     return lastDimensionColOrdinal;
   }
+
+  @Override public String toString() {
```
It's not required. I used it for debug output. Will remove it.

xuchuanyin commented on core/src/main/java/org/apache/carbondata/core/scan/result/BlockletScannedResult.java:
```diff
@@ -369,6 +379,9 @@ public void fillDataChunks() {
     long startTime = System.currentTimeMillis();
     for (int i = 0; i < dimensionColumnPages.length; i++) {
       if (dimensionColumnPages[i][pageCounter] == null && dimRawColumnChunks[i] != null) {
+        // the long string columns is at the end
```
@kumarvishal09
During query, we can get the dimensions that this query requires.
Also we can know how many longStringColumns this query requires.
While writing dimension columns, we write the normal-shortStringColumns first and longStringColumns at last.
So during query, suppose we use `n` dimensions containing `m` longStringColumns.
Then, the first `n-m` columns will be normal-shortStringColumns and the last `m` columns will be longStringColumns.
This line of code can be found below.

kumarvishal09 commented:
@xuchuanyin For supporting String column with more than 32K character we need below changes.
**Create**
1. Support new data type varchar already mentioned by @ravipesala 
**Loading:**
1. Add new encoder . Add this encoder for all the varchar columns in datachunk2 while writing the data to carbondata file. Please check DataChunk2 in carbondata.thrift we are adding encoder for each column
2. Use DirectCompressCodec for compressing the data already code is present in ColumnPage.getLVFlattenedBytePage()
3. Add stats collector for computing max/min for varchar columns implement new class for handling the same
4. No need to add sartkey and endkey for varchar columns,
**Reading**
1. Add new implementation for DimensionDataChunkStore to store INT LV format data.(already handled)
2. Based on encoder present in datachunk2 use DimensionDataChunkStore implementation. Like for dictionary encoder we are creating fixedLengthStoreChunk object 
3. For varchar column just uncompress the data and store the same data with LV format in store (No need to convert LV formatted data to 2D byte array)

**Note:**: We need to handle the same changes for complex data type.
Please take care of backward compatibility :-)
Please let me know for any clarification.
 
@ravipesala @jackylk . Please check if I missed anything.

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6072/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4913/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5074/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6094/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4933/

ravipesala commented:
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5087/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6097/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4935/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5089/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5090/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6240/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5078/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5210/

kumarvishal09 commented:
@xuchuanyin 
1. still use 'long_string_columns' instead of varchar datatype to make it consistent it spark/hive
       Are u facing any problem with varchar??
2. use an integer (previously short) to store the length of bytes content.
     Only for text data type??

xuchuanyin commented:
@ravipesala @kumarvishal09 @jackylk 
Based on @ravipesala 's advice, I changed the name of internal datatype from `text` to `varchar`

Later I'll raise another PR to make it possible for user to directly use `varchar(size)` in DDL statement.

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6259/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5097/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5229/

kumarvishal09 commented on core/src/main/java/org/apache/carbondata/core/datastore/chunk/impl/VariableLengthDimensionColumnPage.java:
```diff
@@ -30,21 +30,19 @@
    */
   public VariableLengthDimensionColumnPage(byte[] dataChunks, int[] invertedIndex,
-      int[] invertedIndexReverse, int numberOfRows) {
+      int[] invertedIndexReverse, int numberOfRows, boolean isVarcharEncoded) {
```
Its better to pass store type it self from the caller instead for boolean

xuchuanyin replied:
OK~

kumarvishal09 commented on core/src/main/java/org/apache/carbondata/core/datastore/chunk/store/impl/safe/SafeVariableLengthDimensionDataChunkStore.java:
```diff
@@ -56,7 +57,8 @@ public SafeVariableLengthDimensionDataChunkStore(boolean isInvertedIndex, int nu
    */
-  @Override public void putArray(final int[] invertedIndex, final int[] invertedIndexReverse,
+  @Override
+  public void putArray(final int[] invertedIndex, final int[] invertedIndexReverse,
```
@xuchuanyin In Case of varchar column I think we will not be able to store complete data in single byte array as it may overflow, in case of short it is fine as maximum it will be of  32000(max number of bytes in each value) * 32000(number of records in page) + 64000(length for each value in a page), so it will be 1024064000 value and as maximum u can create an array of integer max value so it is fine, but in varchar case length of each value can be maximum of Integer max it self so we will not be able to store in single byte array. This handling is required in both data loading and query.

xuchuanyin replied:
@kumarvishal09 yeah, you are right. Considering that scenario will make the code much more complex.
As we all know, a maximum of string/bytearray is about 2GB. I think it is enough for current scenarios. 2GB can support average 67104 bytes per column (2GB/32000) in one column page. 

Having discussed with @jackylk , we have the following conclusions:
1. Add an error message to indicate that the maximum size of one column page is 2GB.
2. We can reduce the row size in the page to support longer characters, for example if the content is 67104*2, user can reduce the row size from 32000 to 16000.

1 will be included in this PR and 2 will be implemented in another PR.
Is that OK? @kumarvishal09

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5265/

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6306/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5144/

xuchuanyin commented:
retest it please

CarbonDataQA commented:
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6325/

CarbonDataQA commented:
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5163/

CarbonDataQA commented:
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6338/

CarbonDataQA commented:
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5176/

ravipesala commented:
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5288/

xuchuanyin commented:
I deleted the origin branch unexpectedly,so I raised another PR #2379
